core/src/test/scala/unit/kafka/server/AbstractFetcherManagerTest.scala

```diff
@@ -210,4 +217,103 @@ class AbstractFetcherManagerTest {
     verify(fetcher).maybeUpdateTopicIds(Set(tp1), topicIds)
     verify(fetcher).maybeUpdateTopicIds(Set(tp2), topicIds)
   }
+
+  @Test
+  def testExpandThreadPool(): Unit = {
+    val topicPartitions = mockTopicPartition(10, 100)
+    val fetcherManager = new AbstractFetcherManager[AbstractFetcherThread]("fetcher-manager", "fetcher-manager", 10) {
+      override def createFetcherThread(fetcherId: Int, sourceBroker: BrokerEndPoint): AbstractFetcherThread = {
+        new TestResizeFetcherThread(sourceBroker)
+      }
+    }
+    try {
+      resizeAndCheckFetcherPartitionDistribution(fetcherManager, topicPartitions, 25)
+    } finally {
+      fetcherManager.closeAllFetchers()
+    }
+  }
+
+  @Test
+  def testShrinkThreadPool(): Unit = {
+    val topicPartitions = mockTopicPartition(10, 100)
+    val fetcherManager = new AbstractFetcherManager[AbstractFetcherThread]("fetcher-manager", "fetcher-manager", 20) {
+      override def createFetcherThread(fetcherId: Int, sourceBroker: BrokerEndPoint): AbstractFetcherThread = {
+        new TestResizeFetcherThread(sourceBroker)
+      }
+    }
+    try {
+      resizeAndCheckFetcherPartitionDistribution(fetcherManager, topicPartitions, 5)
+    } finally {
+      fetcherManager.closeAllFetchers()
+    }
+  }
+
+  def resizeAndCheckFetcherPartitionDistribution(fetcherManager: AbstractFetcherManager[AbstractFetcherThread], topicPartitions: Set[TopicPartition], fetcherNum: Int): Unit = {
+
+    fetcherManager.addFetcherForPartitions(topicPartitions.map(tp => {
+      val brokerId = Utils.abs(tp.topic().hashCode % 5)
+      val brokerEndPoint = new BrokerEndPoint(brokerId, s"kafka-host-$brokerId", 9092)
+      tp -> InitialFetchState(None, brokerEndPoint, 0, 0)
+    }).toMap)
+    fetcherManager.resizeThreadPool(fetcherNum)
+    val ownedPartitions = mutable.Set.empty[TopicPartition]
+    fetcherManager.fetcherThreadMap.forKeyValue { (brokerIdAndFetcherId, fetcherThread) =>
+      val fetcherId = brokerIdAndFetcherId.fetcherId
+      fetcherThread.partitions.foreach { tp =>
+        ownedPartitions += tp
+        assertEquals(fetcherManager.getFetcherId(tp), fetcherId)
+      }
+    }
+    // Verify that all partitions are owned by the fetcher threads.
+    assertEquals(topicPartitions, ownedPartitions)
+  }
+
+  def mockTopicPartition(topicNum: Int, partitionNum: Int): Set[TopicPartition] = {
+    val res = mutable.Set[TopicPartition]()
+    val topicPrefix = "topic_"
+    for (i <- 0 to topicNum) {
+      val topic = topicPrefix + i
+      for (j <- 0 to partitionNum) {
+        res += new TopicPartition(topic, j)
+      }
+    }
+    res.toSet
+  }
+
+  class TestResizeFetcherThread(sourceBroker: BrokerEndPoint)
+    extends AbstractFetcherThread("test-resize-fetcher",
+      clientId = "mock-fetcher",
+      sourceBroker,
+      new FailedPartitions,
+      fetchBackOffMs = 0,
+      brokerTopicStats = new BrokerTopicStats) {
```

Comment:
* nit: Could we also name the parameter for `test-resize-fetcher`?
* nit: Could we align the parameters as well? We can either align all of them on the first one or move the first one on a new line.